ddclient.in

```diff
@@ -2145,6 +2157,157 @@ sub geturl {
     return $reply;
 }
 
+######################################################################
+## geturl -- via curl
+######################################################################
+sub curl_cmd {
+    my @params = @_;
+    my $tmpfile;
+    my %curl_codes = (  ## Subset of error codes from https://curl.haxx.se/docs/manpage.html
+         3 => "URL malformed. The syntax was not correct",
+         5 => "Couldn't resolve proxy. The given proxy host could not be resolved.",
+         6 => "Couldn't resolve host. The given remote host was not resolved.",
+         7 => "Failed to connect to host.",
+        22 => "HTTP page not retrieved. The requested url was not found or returned another error.",
+        28 => "Operation timeout. The specified time-out period was reached according to the conditions.",
+        35 => "SSL connect error. The SSL handshaking failed.",
+        47 => "Too many redirects. When following redirects, curl hit the maximum amount.",
+        51 => "The peer's SSL certificate or SSH MD5 fingerprint was not OK.",
+        58 => "Problem with the local certificate.",
+        60 => "Peer certificate cannot be authenticated with known CA certificates.",
+        67 => "The user name, password, or similar was not accepted and curl failed to log in.",
+        77 => "Problem with reading the SSL CA cert (path? access rights?).",
+        78 => "The resource referenced in the URL does not exist.",
+    );
+
+    return '' if (scalar(@params) == 0); ## no parameters provided
+
+    # debug("CURL:\r\n%s\r\n",$params); ## Uncomment only if required as may expose credentials
+
+    eval "no warnings 'all'; use File::Temp;";
+    if (!$@) {
```

Comment:
I don't like this conditional inclusion because we end up having more code to test and maintain. (It's also not the right way to do conditional inclusion.) If File::Temp isn't available on every system we want to support then we shouldn't use it at all.

Until a compelling use case appears, I'm not willing to jump through hoops to support systems that are missing core modules. Change this to unconditionally import File::Temp (crash out if File::Temp isn't available).

Reply:
> And that is one of the principles you are trying to adhere to—that it will continue to work where it worked before.

No, that's not accurate. If we wanted to maintain compatibility forever we would simply freeze the code and never make any changes. What we're actually trying to do is strike the right balance between code maintainability, features, and support for old systems. See [the Compatibility section of the contributor guide](https://github.com/ddclient/ddclient/blob/master/CONTRIBUTING.md#compatibility) for details.

> the [Astlinux](https://www.astlinux-project.org/) platform does not have it and @abelbeck looked hard at what it would take to add it and found that it drags in a whole pile of other dependencies.

I find that odd, unless Astlinux is missing several core Perl modules. I'd like to know the details.

The cost of working around the lack of File::Temp is too high. Either File::Temp should be added to Astlinux, or this PR should take an approach that doesn't require the creation of a temporary file (e.g., feed the config file to curl via stdin, or use WWW::Curl).

Reply:
> The cost of working around the lack of File::Temp is too high. Either File::Temp should be added to Astlinux, or this PR should take an approach that doesn't require the creation of a temporary file (e.g., feed the config file to curl via stdin, or use WWW::Curl).

I'll let @abelbeck address Astlinux.  If he is able to add it then great.  Or he can look at adding IPC::Open2 or WWW::Curl.  I really don't want to get into the arguments -- I just want to get something done and move on.

I have learned a great deal working on this PR and am very happy with how it looks now.  I think we have a great solution for now and should merge and move on.

Reply:
Uses File::Temp unconditionally now.